### PR TITLE
chore: configuration for private repo auth

### PR DESCRIPTION
**Pull request by yquansah**
Configuration for private repo auth support. If the user provides any one of these parameters we will use it in the `Clone`.

Completes FLI-377

### CODE REVIEW

File: internal/config/storage.go [hunk 1]
```diff
@@ -59,6 +59,21 @@ type Local struct {
 
 // Git contains configuration for referencing a git repository.
 type Git struct {
-	Repository string `json:"repository,omitempty" mapstructure:"repository"`
-	Ref        string `json:"ref,omitempty" mapstructure:"ref"`
+	Repository string    `json:"repository,omitempty" mapstructure:"repository"`
+	Ref        string    `json:"ref,omitempty" mapstructure:"ref"`
+	BasicAuth  BasicAuth `json:"basicAuth,omitempty" mapstructure:"basic_auth"`
+	TokenAuth  TokenAuth `json:"tokenAuth,omitempty" mapstructure:"token_auth"`
```
Comment (markphelps) [inline internal/config/storage.go:65]: would it be possible to make this generic in someway? maybe similar to how we did it for the `AuthenticationConfig` ? https://github.com/flipt-io/flipt/blob/main/internal/config/authentication.go

Or some other way to specify that only one of these should be provided? Its not clear by looking at the struct (and expanding to the env vars/yaml) which version of the auth would take precedence here over the other

Comment (yquansah) [reply internal/config/storage.go:65]: @markphelps Yeah agreed generalizing this structure makes a ton of sense.

File: internal/config/testdata/storage/git_provided.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
 storage:
   type: git
   git:
-    repository: "git@github.com:foo/bar.git"
\ No newline at end of file
+    repository: "git@github.com:foo/bar.git"
```
Comment (yquansah) [inline internal/config/testdata/storage/git_provided.yml:4]: Added new line here that wasn't there previously before.

File: internal/config/storage.go [hunk 3]
```diff
@@ -59,6 +64,83 @@ type Local struct {
...
+// Authentication holds structures for various types of auth we support.
+// Token auth will take priority over Basic auth if both are provided.
+//
+// To make things easier, if there are multiple inputs that a particular auth method needs, and
+// not all inputs are given but only partially, we will return a validation error.
+// (e.g. if username for basic auth is given, and token is also given a validation error will be returned)
+type Authentication struct {
+	Enabled   bool                              `json:"enabled,omitempty" mapstructure:"enabled"`
+	BasicAuth GitAuthentcationMethod[BasicAuth] `json:"basic,omitempty" mapstructure:"basic"`
+	TokenAuth GitAuthentcationMethod[TokenAuth] `json:"token,omitempty" mapstructure:"token"`
```
[the remark this answers is not in the document]
Comment (yquansah) [reply internal/config/storage.go:81]: @GeorgeMac Yeah the design decision makes a lot of sense there. Thanks George!

File: internal/config/storage.go [hunk 4]
```diff
@@ -59,6 +64,64 @@ type Local struct {
...
+				return err
+			}
+		}
+
+		if a.TokenAuth != nil {
+			if err := a.TokenAuth.validate(); err != nil {
+				return err
+			}
+		}
+	}
```
Comment (GeorgeMac) [inline internal/config/storage.go:101]: If we drop enabled, we could just do something like:
```suggestion
	if a.BasicAuth != nil && a.TokenAuth != nil {
		return errors.New("git authentication: cannot enable both basic and token")
	}

	if a.BasicAuth != nil {
		if err := a.BasicAuth.validate(); err != nil {
			return err
		}
	}

	if a.TokenAuth != nil {
		if err := a.TokenAuth.validate(); err != nil {
			return err
		}
	}
```

**Review by GeorgeMac — CHANGES_REQUESTED** [review]
One last thought: otherwise, I think its great.

File: internal/config/storage.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (markphelps) [reply internal/config/storage.go]: yeah you're right. union type is really what we need 👍🏻

File: internal/config/storage.go [hunk 5]
```diff
@@ -59,6 +64,64 @@ type Local struct {
...
+}
+
+// Authentication holds structures for various types of auth we support.
+// Token auth will take priority over Basic auth if both are provided.
+//
+// To make things easier, if there are multiple inputs that a particular auth method needs, and
+// not all inputs are given but only partially, we will return a validation error.
+// (e.g. if username for basic auth is given, and token is also given a validation error will be returned)
+type Authentication struct {
+	Enabled   bool       `json:"enabled,omitempty" mapstructure:"enabled"`
```
[the remark this answers is not in the document]
Comment (yquansah) [reply internal/config/storage.go:79]: @GeorgeMac Yeah agreed. It's most likely unneccesary. I was going for an explicit enabling but probably not necessary.

**Review by GeorgeMac — APPROVED** [review]
Nice one 👍

**Pull request merged**